cpp/FileTransferAgent.cpp

```diff
@@ -103,13 +109,24 @@ void Snowflake::Client::FileTransferAgent::initFileMetadata()
           sourceLocations->at(i));
         break;
       case CommandType::DOWNLOAD:
-        //TODO handle token renew
-        m_FileMetadataInitializer.populateSrcLocDownloadMetadata(
-          sourceLocations->at(i), response.getStageInfo()->getLocation(),
-          m_storageClient);
-        break;
+        {
+          RemoteStorageRequestOutcome outcome =
+            m_FileMetadataInitializer.populateSrcLocDownloadMetadata(
+              sourceLocations->at(i), response.getStageInfo()->getLocation(),
+              m_storageClient, &response.getEncryptionMaterial()->at(i));
+
+          if (outcome == TOKEN_EXPIRED)
+          {
+            sf_log_debug(CXX_LOG_NS,
+                         "Token expired when getting download metadata");
+            this->renewToken(command);
+            i--;
+          }
+          break;
+        }
       default:
         sf_log_error(CXX_LOG_NS, "Invalid command type");
+        throw;
```

Comment:
what exception will be thrown by this?

Reply:
We should throw an SnowflakeException and marked it as internal error since invalid command type should never happen. There are other places that exception need to be thrown. I will revisit this exception handling in a separate commit. (This is done before Kris introduced SnowflakeException class, so I did not throw any explicit exception)